File: java/datastructure/tree/leetcode/SortedArrayToBst.java
package datastructure.tree.leetcode;

/**
 * @author roseduan
 * 将有序数组转换为二叉搜索树
 */
public class SortedArrayToBst {

    public TreeNode sortedArrayToBST(int[] nums) {
        return helper(nums, 0, nums.length - 1);
    }

    private TreeNode helper(int[] nums, int i, int j) {
        if (i > j) {
            return null;
        }

        int mid = (i + j) / 2;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = helper(nums, i, mid - 1);
        root.right = helper(nums, mid + 1, j);
        return root;
    }

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
}

File: go/tree/leetcode/sorted_array_to_bst.go
package leetcode

//将有序数组转换为二叉搜索树

func sortedArrayToBST(nums []int) *TreeNode {
	return convertHelper(nums, 0, len(nums)-1)
}

func convertHelper(nums []int, left, right int) *TreeNode {
	if left > right {
		return nil
	}

	mid := left + (right-left)/2
	root := &TreeNode{Val: nums[mid]}
	root.Left = convertHelper(nums, left, mid-1)
	root.Right = convertHelper(nums, mid+1, right)
	return root
}
